docs: Describe the process of upgrading settings.py files.

alexmv opened the pull request:
Fixes zulip#17782.

Also included is a second commit which attempts to find the changes you made from the template you most recently used.  Here's what that looks like:
```
root@alexmv-prod:~# ~zulip/deployments/current/scripts/minimal-template-diff
Reading current configuration from /etc/zulip/settings.py...
Fetching old versions of setting templates from Github...
  - 3.4
  - 3.3
  - 3.2
  - 3.1
  - 3.0
  - 2.1.7
  - 2.1.6
  - 2.1.5
  - 2.1.4
  - 2.1.3
  - 2.1.2
  - 2.1.1
  - 2.1.0
  - 2.0.8
Computing minimal difference...
Most likely version: 3.2

--- zulip-server-{version}/zproject/prod_settings_template.py
+++ /etc/zulip/settings.py
@@ -1,4 +1,6 @@
 from typing import Any, Dict, Tuple
+
+from zproject.config import get_secret

 ################################################################
 # Zulip Server settings.
@@ -27,7 +29,7 @@
 # support@example.com is totally reasonable, as is admin@example.com.
 # Do not put a display name; e.g. 'support@example.com', not
 # 'Zulip Support <support@example.com>'.
-ZULIP_ADMINISTRATOR = 'zulip-admin@example.com'
+ZULIP_ADMINISTRATOR = 'alexmv@zulip.com'

 # The user-accessible Zulip hostname for this installation, e.g.
 # zulip.example.com.  This should match what users will put in their
@@ -36,7 +38,7 @@
 #
 # If you need to access the server on a specific port, you should set
 # EXTERNAL_HOST to e.g. zulip.example.com:1234 here.
-EXTERNAL_HOST = 'zulip.example.com'
+EXTERNAL_HOST = 'alexmv-prod.zulipdev.org'

 # Alternative hostnames.  A comma-separated list of strings
 # representing the host/domain names that your users can enter in
@@ -69,18 +71,15 @@
 # advice for troubleshooting, see the Zulip documentation:
 #   https://zulip.readthedocs.io/en/latest/production/email.html

-# EMAIL_HOST and EMAIL_HOST_USER are generally required.
-#EMAIL_HOST = 'smtp.example.com'
-#EMAIL_HOST_USER = ''
-
-# Passwords and secrets are not stored in this file.  The password
-# for user EMAIL_HOST_USER goes in `/etc/zulip/zulip-secrets.conf`.
-# In that file, set `email_password`.  For example:
-#   email_password = abcd1234
-
-# EMAIL_USE_TLS and EMAIL_PORT are required for most SMTP providers.
-#EMAIL_USE_TLS = True
-#EMAIL_PORT = 587
+#EMAIL_BACKEND = 'django_ses.SESBackend'
+#AWS_SES_ACCESS_KEY_ID = get_secret("email_username")
+#AWS_SES_SECRET_ACCESS_KEY = get_secret("email_password")
+
+# Send using SES; see https://docs.aws.amazon.com/general/latest/gr/ses.html for hosts.
+EMAIL_HOST = "email-smtp.us-east-1.amazonaws.com"
+EMAIL_HOST_USER = get_secret("email_username")
+EMAIL_USE_TLS = True
+EMAIL_PORT = 587

 # The noreply address to be used as the sender for certain generated
 # emails.  Messages sent to this address could contain sensitive user
@@ -362,9 +361,9 @@

 # Controls whether or not error reports (tracebacks) are emailed to the
 # server administrators.
-#ERROR_REPORTING = True
+ERROR_REPORTING = True
 # For frontend (JavaScript) tracebacks
-#BROWSER_ERROR_REPORTING = False
+BROWSER_ERROR_REPORTING = True

 # If True, each log message in the server logs will identify the
 # Python module where it came from.  Useful for tracking down a
@@ -437,9 +436,9 @@
 # Valid values for REMOTE_POSTGRES_SSLMODE are documented in the
 # "SSL Mode Descriptions" table in
 #   https://www.postgresql.org/docs/9.5/static/libpq-ssl.html
-#REMOTE_POSTGRES_HOST = 'dbserver.example.com'
+#REMOTE_POSTGRES_HOST = 'localhost'
 #REMOTE_POSTGRES_PORT = '5432'
-#REMOTE_POSTGRES_SSLMODE = 'require'
+#REMOTE_POSTGRES_SSLMODE = 'disable'

 # If you want to set a Terms of Service for your server, set the path
 # to your markdown file, and uncomment the following line.
@@ -643,12 +642,12 @@
 # thumbor on a different server, you can configure that by setting
 # THUMBOR_URL here.  Setting THUMBOR_URL='' will let Zulip server know that
 # thumbor is not running or configured.
-#THUMBOR_URL = 'http://127.0.0.1:9995'
+THUMBOR_URL = 'http://127.0.0.1:9995'
 #
 # This setting controls whether images shown in Zulip's inline image
 # previews should be thumbnailed by thumbor, which saves bandwidth but
 # can modify the image's appearance.
-#THUMBNAIL_IMAGES = True
+THUMBNAIL_IMAGES = False

 # Controls the Jitsi Meet video call integration.  By default, the
 # integration uses the SaaS meet.jit.si server.  You can specify
```

One problem is that the unauth'd Github API is [rate-limited to 60 requests per hour](https://docs.github.com/en/rest/overview/resources-in-the-rest-api#rate-limiting) so you can only run the tool 4 times per hour, and that number will go down as we make more releases.

If this seems plausibly useful, I'll amend the second commit with instructions pointing at it.

timabbott commented on docs/production/settings.md:
```diff
@@ -79,17 +83,19 @@ configuration along with your other Zulip server configuration.
+`/etc/zulip/settings.py`; you can review [the latest version of the
+settings.py template][settings-py-template], and we recommend
+[carefully updating your `/etc/zulip/settings.py`][update-settings] to
+fold in new configuration settings and documentation.
```
Maybe "to fold in the inline comment documentation for new settings"?

I want to make clear that it's not a functional step.

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
+   there, making it match the old value in `/etc/zulip/settings.py`.
+
+   If there are settings which you cannot find documented in
+   `/etc/zulip/settings-new.py`, check the [change log][changelog] to see
```
I think "changelog" is usually one word?

timabbott commented on docs/subsystems/settings.md:
```diff
@@ -57,7 +57,10 @@ In a production environment, we have:
+  `zproject/default_settings.py`, and then override it here.  As this
+  is the main documentation for Zulip settings, we recommend [carefully
+  updating your `/etc/zulip/settings.py` every major
+  release](../production/upgrade-or-modify.html#updating-configuration-files).
```
Should we frame this as an explanation for Zulip developers about this recommendation we make to production installations (i.e. not using "your" here)?

alexmv replied:
:nod:  Makes sense, since it's a different audience than the other locations.

alexmv commented:
Did you have opinions about the utility of the tool in the second commit?

timabbott commented:
The tool seems useful to me.  The rate-limit seems like a minor problem.

We could replace the `github.com` queries with hosting copies of those files on zulip.org/dist somewhere; it seems not hard to extend our release script to extract that one file from the release tarballs and putting it in a directory.

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
 See also our documentation on [upgrading
 docker-zulip](https://github.com/zulip/docker-zulip#upgrading-from-a-git-repository).
 
+## Updating configuration files
```
Is this the right place to put this?  I'm not sure if it's important to have "troubleshooting" more close to the actual upgrade instructions.  I guess we already link to there from the non-Git instructions, and upgrading from Git is relatively rare.

Another theory is that we could put this just before "Upgrading the operating system", and then link to it from https://zulip.readthedocs.io/en/latest/overview/changelog.html#upgrade-notes (which we encourage folks to read whenever looking at doing an upgrade).

This is a step that one only needs to do occasionally, and it's OK if folks forget; we're basically just providing updated inline documentation.

timabbott replied:
I think we can try this out and iterate from here; it's not hard to backport documentation changes if we decide to move this.

(I do worry a bit that the heading here should be instead "Updating settings.py comment documentation" -- so that it's clear from scanning the headings that nothing bad will happen if you don't update the configuration files atomically with the update itself.

alexmv replied:
Yeah.  I think the other heading which is confusing here is "Preserving local changes to configuration files" which is referring to things like nginx and postgres, but sits weirdly with "Updating configuration files" right near it.

alexmv replied:
Reworded both of these headings.

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
+
+1. Ensure you have a
+   [backup](../production/export-and-import.html#backups) of your
+   `/etc/zulip/settings.py` file.
```
Should we just recommend a `cp -a` command here?  They don't need a full system backup for this.  Maybe:
"If you don't have [backups](), make a copy of your `/etc/zulip` directory:
```
cp -a /etc/zulip/settings.py ~/zulip-settings-backup.py
```

timabbott replied:
Yeah, I think advertising backups here is great, but for this task, `cp -a` really is good enough, since it's just a temporary copy to save you from editing mistakes.  I like the language you landed on.

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
+
+1. Copy
+   `/home/zulip/deployments/current/zproject/prod_settings_template.py`
+   to `/etc/zulip/settings-new.py`.
```
Should we just give a `cp` command, maybe collapsing this with the previous step?

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
+   if they have been removed.
+
+1. Move `/etc/zulip/settings-new.py` to `/etc/zulip/settings.py`,
+   overwriting the old configuration.
```
Especially with your new tool, I think we can reasonably recommend that they run the tool, look at the diffs, and decide either port in changes as described here or just add new entries (since we may have releases where the churn is just a new block), rather than only suggesting this approach (which is surely the only sane one for 4.0, since we redid the whitespace).

alexmv replied:
Knowing which entries are "new" may be a little hard.  I guess the tool can diff the old template to the new template, and spit that out?  That seems like it might make the output confusing to parse.

timabbott replied:
Yeah, and I guess especially with this update (Where we reordered things), diffing the only and new templates is probably not going to be that helpful.

alexmv replied:
Very not-helpful.  But for many upgrades, it'll probably be reasonable.

I guess we can address that in 5.0 or whenever we have config changes that would be amenable to this. :)

timabbott commented on docs/production/upgrade-or-modify.md:
```diff
@@ -102,6 +102,43 @@ git_repo_url = https://github.com/zulip/zulip.git
+
+1. Go through each variable set in `/etc/zulip/settings.py`, updating
+   `/etc/zulip/settings-new.py`, wherever that setting is found in
+   there, making it match the old value in `/etc/zulip/settings.py`.
```
May be worth saying explicitly we recommend opening both files in an editor, and for each variable set in their `settings.py`, find its section in the new one and copy it to there?

We could also suggest `grep '^[^#]' /etc/zulip/settings.py` here.

timabbott commented on docs/production/settings.md:
```diff
@@ -19,6 +19,10 @@ restart the server with the following command:
 
+As this is the main documentation for Zulip settings, we recommend
+[carefully updating your `/etc/zulip/settings.py` every major
+release](../production/upgrade-or-modify.html#updating-configuration-files).
```
Maybe this one should provide a bit more context, since it's explaining how settings work overall:
``` quote
`/etc/zulip/settings.py` has useful documentation for dozens of minor settings in its comments. If you've upgraded from an old version of Zulip, your `settings.py` file will not have comments for any settings added since the Zulip release you originally installed; for this reason, we recommend [updating `settings.py`]() after major releases.
```

timabbott replied:
Or maybe just deciding that the existing "miscellaneous" section you have below is good enough, since my suggestion here is basically the exact same text you have there.

timabbott replied:
OK here's a proposal:
* We move "miscellaneous server settings" content to sit here (since its second half just being a new start of "Specific settings" or something).

timabbott replied:
Yeah, I think this works well.

timabbott commented:
This lgtm aside from:
* The heading issue in https://github.com/zulip/zulip/pull/18239#discussion_r621318085.
* Whether we want to setup a directory in zulip.com/dist with copies of `prod_settings_template.py` to avoid the GitHub API rate limits (which could reasonably be a follow-up project to let us merge this).

alexmv commented:
> This lgtm aside from:
> 
> * The heading issue in [#18239 (comment)](https://github.com/zulip/zulip/pull/18239#discussion_r621318085).

Adjusted both headings (see above) and also changed the wording in the couple places it's linked to make more explicit that it's just documentation updates.

> * Whether we want to setup a directory in zulip.com/dist with copies of `prod_settings_template.py` to avoid the GitHub API rate limits (which could reasonably be a follow-up project to let us merge this).

I swapped it over to using `raw.githubusercontent.com` for the actual file contents, so it's just fetching the tags (once per call) that hits the API endpoint -- which means now you can run it 60 times per hour.  I think that's reasonable for a first pass, and not worth the time to set up more infra to deal with this right now.

alexmv commented:
Updated commit message for the tool to doc the rate-limit.

timabbott commented on docs/production/upgrade-or-modify.md:
````diff
@@ -133,6 +133,14 @@ suggest using that updated template to update
+   using, and the differences that your file has from that:
+
+   ```
+   /home/zulip/deployments/current/scripts/minimal-template-diff
````
Let's move this to `scripts/setup/` -- we try to keep `scripts/` for things that have good `--help` and are things we expect users to run routinely.

It may also want a clearer name; maybe `compare-settings-to-template` or something?

timabbott commented:
OK, that's great -- posted one last comment about the path for the script.

timabbott commented:
(And I went ahead and merged the first commit as bc9d20eaa2637a45581f1f4653fc0f9292d64d9d)

timabbott commented:
Merged, thanks @alexmv!
